File: apps/ingestor/internal/event/hash.go
package event

import (
	"crypto/sha256"
	"encoding/binary"
)

// hash64 returns a deterministic 64-bit hash of the input, suitable for a
// ClickHouse UInt64 column (first 8 bytes of SHA-256, big-endian).
func hash64(input string) uint64 {
	sum := sha256.Sum256([]byte(input))
	return binary.BigEndian.Uint64(sum[:8])
}

File: apps/ingestor/src/lib/hash.ts
import { createHash } from 'node:crypto';

/**
 * Deterministic 64-bit hash of a string, returned as a bigint suitable for a
 * ClickHouse UInt64 column. Takes the first 8 bytes of SHA-256.
 */
export function hash64(input: string): bigint {
	const digest = createHash('sha256').update(input).digest();
	return digest.readBigUInt64BE(0);
}
